src/tmpfiles/tmpfiles.c

```diff
@@ -554,6 +554,58 @@ static nsec_t load_statx_timestamp_nsec(const struct statx_timestamp *ts) {
         return ts->tv_sec * NSEC_PER_SEC + ts->tv_nsec;
 }
 
+static int opendir_and_stat(
+                const char *path,
+                DIR **ret,
+                nsec_t *ret_atime_nsec,
+                nsec_t *ret_mtime_nsec,
+                dev_t *ret_rootdev_major,
+                dev_t *ret_rootdev_minor,
+                bool *ret_mountpoint) {
+
+        _cleanup_closedir_ DIR *d = NULL;
+        STRUCT_STATX_DEFINE(sx);
+        int r;
+
+        assert(path);
+        assert(ret);
+        assert(ret_atime_nsec);
+        assert(ret_mtime_nsec);
+        assert(ret_rootdev_major);
+        assert(ret_rootdev_minor);
+        assert(ret_mountpoint);
+
+        d = opendir_nomod(path);
+        if (!d)
+                return log_full_errno(IN_SET(errno, ENOENT, ENOTDIR) ? LOG_DEBUG : LOG_ERR,
+                                      errno, "Failed to open directory %s: %m", path);
+
+        r = statx_fallback(dirfd(d), "", AT_EMPTY_PATH, STATX_MODE|STATX_INO|STATX_ATIME|STATX_MTIME, &sx);
+        if (r < 0)
+                return log_error_errno(r, "statx(%s) failed: %m", path);
+
+        if (FLAGS_SET(sx.stx_attributes_mask, STATX_ATTR_MOUNT_ROOT))
+                *ret_mountpoint = FLAGS_SET(sx.stx_attributes, STATX_ATTR_MOUNT_ROOT);
+        else {
+                struct stat ps;
+
+                if (fstatat(dirfd(d), "..", &ps, AT_SYMLINK_NOFOLLOW) != 0)
+                        return log_error_errno(errno, "stat(%s/..) failed: %m", path);
+
+                *ret_mountpoint =
+                        sx.stx_dev_major != major(ps.st_dev) ||
+                        sx.stx_dev_minor != minor(ps.st_dev) ||
+                        sx.stx_ino != ps.st_ino;
```

Comment:
hmm, i don't grok the last part of this comparison? this sounds broken? (i am ware this is a pre-existing issue).

seems like a candidate for replacement by `statx_mount_same()`